File: 00_Stack/NodeJS/stack.js
// Stack implementation nodejs Version 1 (functional way)

// // Create stack
// var createStack = function() {
//     const stack = [];
//     return stack;
// }

// // Creating an empty stack
// var checkEmpty = function(stack) {
//     return stack.lenth == 0;
// }

// // Adding items into the stack
// var push = function(stack, item) {
//     stack.push(item);
//     console.log("pushed item:" + item)
// }

// // Removing an element from the stack
// var pop = function(stack) {
//     if (checkEmpty(stack)) {
//         return "stack is empty"
//     }

//     return stack[stack.length -1];
// }

// var stack = createStack();
// push(stack, new String('one'));
// push(stack, new String('two'));
// push(stack, new String('three'));
// push(stack, new String('four'));

// let popped = pop(stack);

// console.log(popped);

// Stack implementation nodejs Version 2 (like object )
// const StdStack = {
//     // __constructor() {
//     //     return [];
//     // },
//     createStack: () => [],
//     checkEmpty: (stack) =>  stack.lenth == 0,
//     push: function(stack, item) {
//         stack.push(item);
//         console.log("pushed item:" + item);
//     },
//     pop: function(stack) {
//         if ( true === this.checkEmpty(stack)) {
//             return "stack is empty";
//         }

//         return stack[stack.length -1];
//     }
//   };

// const stack1    = StdStack.createStack();
// const strOne    = new String('one');
// const strTwo    = new String('two');
// const id        =  Symbol.for("12323");

// console.log(id);
// let clone = Object.assign({}, user);
// StdStack.push(new String('one'))
// StdStack.push(stack1, new String('one'))
// StdStack.push(stack1, new String('one'))
// StdStack.push(stack1, new String('one'))
// StdStack.push(stack1, new Symbol('id'))
// console.info(stack1);
// StdStack.push(new String('one'))
// console.log();
// StdStack.push(new String('one'))
// console.log();

// Stack implementation nodejs Version 3 (like function )

// class has contructor
class StdStack {

    /**
     * @param <Number> size 
     */
    constructor(size = 5) {
        this.size = new Number(size);
        this.stack = [];
    }

    /**
     * @return bool 
     */
    isEmpty() { 
        return this.stack.lenth == 0;
    }

    /**
     * @param <Number> item 
     */
    push(item) {

        if ( true === this.isFull()) {
            throw new Error('Stack is full');
        }

        this.stack.push(item);
        console.log("pushed item:" + item);
    } 
 
    /**
     * @returns <any>
     */
    pop() {
        if ( true === this.isEmpty()) {
            return "stack is empty";
        }

        return this.stack.pop();
    }

    /**
     * @returns bool 
     */
    isFull() {
        if (this.stack.length < this.size) {
            console.log('Stack is full');
            return false;
        }
        
        return true;
    }

    /**
     * @returns Element content;
     */
    peek() {
        return this.stack[this.stack.lenth];
    }
}
var stack = new StdStack(2);
stack.push(new String('one'));
stack.push(new String('two'));
stack.push(new String('three'));
// stack.push(new String('four'));

// let clone = Object.assign({}, user);
// StdStack.push(new String('one'))
// console.info(stack1);
// StdStack.push(new String('one'))
// console.log();
// StdStack.push(new String('one'))
// console.log();

File: 00_Stack/C/stack.c
// Stack implementation in C

#include <stdio.h>
#include <stdlib.h>

#define MAX 2

int count = 0;

// Creating a stack
struct Stack {
    int items[MAX];
    int top;
};
typedef struct Stack stackv;

void createEmptyStack(stackv *st) {
    st->top = -1;
}

// Check if the stack is full
int isfull(stackv *st) {
    if (st->top == MAX - 1)
        return 1;
    else 
        return 0;
}

// Check if the stack is empty
int isempty(stackv *st) {
    if (st->top == -1)
        return 1;
    else
        return 0;
}

// Add elements into stack
void push(stackv *st, int newitem) {
    if (isfull(st)) {
        printf("STACK FULL");
    } else {
        st->top++;
        st->items[st->top] = newitem;
    }
    count++;
}

// Remove element from stack
void pop(stackv *st) {
    if (isempty(st)) {
        printf("\n STACK EMPTY \n");
    } else {
        printf("Item popped= %d", st->items[st->top]);
        st->top--;
    }
    count--;
    printf("\n");
}

void printStack(stackv *st) {
    printf("Stack: ");
    for (int i = 0; i < count; i++) {
        printf("%d ", st->items[i]);
    }
    printf("\n");
}

// Launch code
int main() {
    // int ch;
    stackv *st = (stackv *)malloc(sizeof(st));

    createEmptyStack(st);

    push(st, 1);
    push(st, 11);
    push(st, 111);
    push(st, 1111);

    printStack(st);

    pop(st);

    printf("\nAfter popping out\n");
    printStack(st);
}